File: python/insertion.py
def insertion_sort(arr):
    n = len(arr)

    for i in range(1, n):
        key = arr[i]
        j = i - 1

        print(f"\nPass {i}:")
        print(f"Ambil key = {key}")

        # Geser elemen yang lebih besar dari key
        while j >= 0 and arr[j] > key:
            print(f"Bandingkan {arr[j]} > {key} -> Geser")
            arr[j + 1] = arr[j]
            j -= 1
            print(" ", arr)

        # Masukkan key ke posisi yang benar
        arr[j + 1] = key
        print(f"Masukkan key di posisi {j+1}")
        print(" ", arr)

    return arr


# Input
n = int(input("Masukkan jumlah data: "))

while True:
    data = list(map(int, input("Masukkan data: ").split()))
    if len(data) != n:
        print(f"Harus {n} data!")
    else:
        break

# Proses
hasil = insertion_sort(data)

# Output
print("\nHasil akhir:", hasil)

File: c++/insertion.cpp
#include <iostream>
using namespace std;

void tampil(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

void insertionSort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        cout << "\nPass " << i << ":" << endl;
        cout << "Ambil key = " << key << endl;

        while (j >= 0 && arr[j] > key) {
            cout << "Bandingkan " << arr[j] << " > " << key << " -> Geser" << endl;
            
            arr[j + 1] = arr[j];
            j--;

            tampil(arr, n);
        }

        arr[j + 1] = key;

        cout << "Masukkan key di posisi " << j + 1 << endl;
        tampil(arr, n);
    }
}

int main() {
    int n;
    cout << "Masukkan jumlah data: ";
    cin >> n;

    int arr[100];
    cout << "Masukkan data (pisah spasi): ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    insertionSort(arr, n);

    cout << "\nHasil akhir: ";
    tampil(arr, n);

    return 0;
}
